### PR TITLE
[repl][completion-epc] added completion support for emacs.

### PR DESCRIPTION
**Pull request by otadmor**
This uses the generic completion-epc.
Tested on python3, python2, completion-epc server mode and completion-epc client mode.

### CODE REVIEW

**Comment by oleavr** [comment]
Cool!! Could you update the formatting to match [PEP-8](https://www.python.org/dev/peps/pep-0008/)? Specifically pay attention to spaces, w/blank lines, and using `snake_case` instead of `camelCase` APIs where available. Also it would be great to move the logic down to before the part with `input_impl`. (Code should follow [Newspaper Code Structure](https://kentcdodds.com/blog/newspaper-code-structure).)

**Comment by otadmor** [comment]
Sure.

**Comment by otadmor** [comment]
I did commit amend and now the new code is according to pep8 (i did not changed the rest of repl to fit pep8 as it is out of scope). I haven't see anything in camelCase but the setDeamon which dont have set_deamon afaik. Could you refer me to a place where I did camelCase I think I missed it?

[code]
>>> import threading
>>> t=threading.Thread()
>>> t.setDaemon
<bound method Thread.setDaemon of <Thread(Thread-1, initial)>>
>>> t.set_daemon
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Thread' object has no attribute 'set_daemon'
[/code]

File: frida_tools/repl.py [hunk 1]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
             self._result = (None, KeyboardInterrupt())
             self._cond.notify()
```
Comment (oleavr) [inline frida_tools/repl.py]: Let's add one more blank line here.

File: frida_tools/repl.py [hunk 2]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
 
+if os.environ.get("TERM", '') == "dumb":
```
Comment (oleavr) [inline frida_tools/repl.py]: The convention is to use double-quoted strings for strings, and single-quoted for enum/magic values:

```suggestion
if os.environ.get("TERM", "") == 'dumb':
```

(In hindsight this is a bit nutty, but let's try to stick to it for now.)

File: frida_tools/repl.py [hunk 3]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+        from epc.server import EPCServer
+        from epc.client import EPCClient
+        from collections import namedtuple
+    except ImportError:
+        def start_completion_thread(repl, epc_port=None):
+            # do nothing when cannot import epc module.
+            pass
+    else:
+        class EPCCompletionServer(EPCServer):
+            def __init__(self, address='localhost', port=0, *args, **kargs):
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
            def __init__(self, address="localhost", port=0, *args, **kargs):
```

File: frida_tools/repl.py [hunk 4]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+                EPCServer.__init__(self, (address, port), *args, **kargs)
+
+                def complete(*cargs, **ckargs):
+                    return self.complete(*cargs, **ckargs)
+                self.register_function(complete)
+
+            def print_port(self, stream=None):
+                if stream is None:
+                    stream = sys.stdout
+                stream.write('___EPCCompletionServer_PORT=')
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
                stream.write("___EPCCompletionServer_PORT=")
```

File: frida_tools/repl.py [hunk 5]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+                EPCServer.print_port(self, stream)
+
+        class EPCCompletionClient(EPCClient):
+            def __init__(self, address='localhost', port=None, *args, **kargs):
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
            def __init__(self, address="localhost", port=None, *args, **kargs):
```

File: frida_tools/repl.py [hunk 6]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+                if port is not None:
+                    args = ((address, port), ) + args
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
                    args = ((address, port),) + args
```

File: frida_tools/repl.py [hunk 7]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+                EPCClient.__init__(self, *args, **kargs)
+
+                def complete(*cargs, **ckargs):
+                    return self.complete(*cargs, **ckargs)
+                self.register_function(complete)
+
+        EpcDocument = namedtuple('Document', ['text_before_cursor', ])
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
        EpcDocument = namedtuple('Document', ['text_before_cursor',])
```

File: frida_tools/repl.py [hunk 8]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+
+        class ReplEPCCompletion(object):
+            def __init__(self, repl, *args, **kargs):
+                self._repl = repl
+
+            def complete(self, *to_complete):
+                to_complete = ''.join(to_complete)
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
                to_complete = "".join(to_complete)
```

File: frida_tools/repl.py [hunk 9]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+                prefix = '.'.join(to_complete.split(".")[:-1])
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
                prefix = ".".join(to_complete.split(".")[:-1])
```

File: frida_tools/repl.py [hunk 10]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+
+        def start_completion_thread(repl, epc_port=None):
+            if epc_port is None:
+                epc_port = os.environ.get("EPC_COMPLETION_SERVER_PORT", None)
+            if epc_port is not None:
+                epc_port = int(epc_port)
+                rpc_complete = ReplEPCCompletionClient(repl, port=epc_port)
+                rpc_complete_thread = threading.Thread(
+                    target=rpc_complete.connect,
+                    name='PythonModeEPCCompletion',
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
                    name="PythonModeEPCCompletion",
```

File: frida_tools/repl.py [hunk 11]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+                    kwargs={'socket_or_address': ('localhost', epc_port)})
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
                    kwargs={'socket_or_address': ("localhost", epc_port)})
```

File: frida_tools/repl.py [hunk 12]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+            else:
+                rpc_complete = ReplEPCCompletionServer(repl)
+                rpc_complete.print_port()  # needed for Emacs client
+                rpc_complete_thread = threading.Thread(
+                    target=rpc_complete.serve_forever,
+                    name='PythonModeEPCCompletion')
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
                    name="PythonModeEPCCompletion")
```

File: frida_tools/repl.py [hunk 13]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
             self._cond.notify()
 
+if os.environ.get("TERM", '') == "dumb":
+    try:
+        from epc.server import EPCServer
+        from epc.client import EPCClient
+        from collections import namedtuple
+    except ImportError:
+        def start_completion_thread(repl, epc_port=None):
+            # do nothing when cannot import epc module.
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
            # Do nothing when we cannot import the EPC module.
```

File: frida_tools/repl.py [hunk 14]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+                rpc_complete.print_port()  # needed for Emacs client
+                rpc_complete_thread = threading.Thread(
+                    target=rpc_complete.serve_forever,
+                    name='PythonModeEPCCompletion')
+            rpc_complete_thread.setDaemon(True)
+            rpc_complete_thread.start()
+            return rpc_complete_thread
+else:
+    def start_completion_thread(repl, epc_port=None):
+        # do nothing as completion-epc is not needed when not running in emacs.
```
Comment (oleavr) [inline frida_tools/repl.py]:
```suggestion
        # Do nothing as completion-epc is not needed when not running in Emacs.
```

File: frida_tools/repl.py [hunk 15]
```diff
@@ -757,6 +760,92 @@ def _cancel_line(self):
...
+                    target=rpc_complete.connect,
+                    name='PythonModeEPCCompletion',
+                    kwargs={'socket_or_address': ('localhost', epc_port)})
+            else:
+                rpc_complete = ReplEPCCompletionServer(repl)
+                rpc_complete.print_port()  # needed for Emacs client
+                rpc_complete_thread = threading.Thread(
+                    target=rpc_complete.serve_forever,
+                    name='PythonModeEPCCompletion')
+            rpc_complete_thread.setDaemon(True)
```
Comment (oleavr) [inline frida_tools/repl.py]: Let's use the modern API instead:

```suggestion
            rpc_complete_thread.daemon = True
```

That way we also avoid the camelCase method breaking the style. :-)

File: frida_tools/repl.py (hunk 1 above)
Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion


```

File: frida_tools/repl.py [hunk 16]
```diff
@@ -3,7 +3,7 @@
 
 import signal
 import threading
-
+import os
```
Comment (oleavr) [inline frida_tools/repl.py]: Let's move this up so they are in alphabetic order. Also we need to keep two blank lines between the imports and `main`.

File: frida_tools/repl.py [hunk 17]
```diff
@@ -3,7 +3,7 @@
```
Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion

import os
```

File: frida_tools/repl.py (hunk 16 above)
Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion
```

Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion
```

File: frida_tools/repl.py (hunk 17 above)
Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion

import os```

File: frida_tools/repl.py [hunk 18]
```diff
@@ -1,10 +1,10 @@
 # -*- coding: utf-8 -*-
 from __future__ import unicode_literals, print_function
 
+import os```
```
Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion
import os

File: frida_tools/repl.py [hunk 19]
```diff
@@ -1,10 +1,10 @@
 # -*- coding: utf-8 -*-
 from __future__ import unicode_literals, print_function
 
+import os
 import signal
 import threading
 
-
```
Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion

Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion

File: frida_tools/repl.py [hunk 20]
```diff
@@ -1,10 +1,10 @@
 # -*- coding: utf-8 -*-
 from __future__ import unicode_literals, print_function
 
+import os
 import signal
 import threading
```
Comment (otadmor) [inline frida_tools/repl.py]:
```suggestion

**Comment by otadmor** [comment]
Thanks!

**Pull request merged**